components/brave_wallet/browser/zcash/zcash_scan_blocks_task.cc

```diff
@@ -0,0 +1,170 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_scan_blocks_task.h"
+
+#include <algorithm>
+#include <utility>
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_wallet_service.h"
+
+namespace brave_wallet {
+
+namespace {
+constexpr uint32_t kBatchSize = 1024;
+}  // namespace
+
+ZCashScanBlocksTask::ZCashScanBlocksTask(
+    ZCashActionContext& context,
+    ZCashShieldSyncService::OrchardBlockScannerProxy& scanner,
+    ZCashScanBlocksTaskObserver observer,
+    std::optional<uint32_t> to)
+    : context_(context),
+      scanner_(scanner),
+      observer_(std::move(observer)),
+      to_(to) {}
+
+ZCashScanBlocksTask::~ZCashScanBlocksTask() = default;
+
+void ZCashScanBlocksTask::Start() {
+  CHECK(!started_);
+  started_ = true;
+  ScheduleWorkOnTask();
+}
+
+void ZCashScanBlocksTask::ScheduleWorkOnTask() {
+  base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+      FROM_HERE, base::BindOnce(&ZCashScanBlocksTask::WorkOnTask,
+                                weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashScanBlocksTask::WorkOnTask() {
+  if (error_) {
+    observer_.Run(base::unexpected(*error_));
+    return;
+  }
+
+  if (!account_meta_) {
+    GetAccountMeta();
+    return;
+  }
+
+  if (!chain_tip_block_) {
+    GetChainTip();
+    return;
+  }
+
+  if (!scan_ranges_) {
+    PrepareScanRanges();
+    return;
+  }
+
+  if (!scan_ranges_->empty()) {
+    ScanRanges();
+    return;
+  }
+}
+
+void ZCashScanBlocksTask::PrepareScanRanges() {
+  CHECK(account_meta_);
+  uint32_t from = account_meta_->latest_scanned_block_id
+                      ? account_meta_->latest_scanned_block_id.value() + 1
+                      : account_meta_->account_birthday;
+  if (to_ && (chain_tip_block_.value() < to_.value() || to_.value() < from)) {
+    error_ = ZCashShieldSyncService::Error{
+        ZCashShieldSyncService::ErrorCode::kFailedToUpdateChainTip,
+        "Scan range error"};
+    ScheduleWorkOnTask();
+    return;
+  }
+  uint32_t to = to_.value_or(chain_tip_block_.value());
+
+  start_block_ = from;
+  end_block_ = to;
+
+  initial_ranges_count_ =
+      std::ceil(static_cast<double>((to - from + 1)) / kBatchSize);
+  scan_ranges_ = std::deque<ScanRange>();
```

Comment:
Can it be rewritten somehow so it's easier to track if always `to >= form`?

Reply:
done